File: landa/landa_sales/delivery_note/delivery_note.js
frappe.ui.form.on('Delivery Note',  {
    refresh: function (frm) {
        if (frm.doc.is_return) {
            // some items cannot be returned
            frm.set_query("item_code", "items", function () {
                return {
                    query: "erpnext.controllers.queries.item_query",
                    filters: {
                        'cannot_be_returned': 0,
                        'is_sales_item': 1
                    }
                };
            });
        }
    },
    before_save: function (frm) {
        frm.doc.items = frm.doc.items.filter(function (value) {
            return value.qty != 0;
        });
    },
    customer: function (frm) {
        frm.trigger('prefill_delivery_items');
    },
    year_of_settlement: function (frm) {
        frm.trigger('prefill_delivery_items');
    },
    prefill_delivery_items: function (frm) {
        if (frm.doc.customer != null && frm.doc.year_of_settlement != null) {
            frappe.call({
                method: "landa.landa_sales.delivery_note.delivery_note.get_items",
                args: {
                    'year': frm.doc.year_of_settlement
                },
                callback: function (r) {
                    frm.clear_table("items")
                    frm.refresh_field("items")
                    for (const item of r.message) {
                        const row = frm.add_child("items")
                        row.item_code = item.item_code
                        row.item_name = item.item_name
                        row.cannot_be_returned = item.cannot_be_returned
                        row.description = item.description
                        row.uom = item.uom
                        row.uom_factor = item.uom_factor
                        row.qty = item.qty
                        row.rate = item.rate
                    }
                    frm.refresh_field("items");
                }
            })
        }
        else {
            frm.clear_table("items")
            frm.refresh_field("items")
        }
    }
});

File: landa/landa_sales/delivery_note/delivery_note.py
import frappe
from frappe import _
from erpnext.stock.doctype.delivery_note.delivery_note import make_sales_invoice


def validate(delivery_note, event):
	"""Validate that only returnable items are returned."""
	if delivery_note.is_return:
		for item in delivery_note.items:
			if frappe.db.get_value('Item', item.item_code, 'cannot_be_returned'):
				frappe.throw(_('Item {} cannot be returned').format(item.item_name))
				return


def on_submit(delivery_note, event):
	"""Auto-create Sales Invoice for Delivery Note."""
	if delivery_note.get('create_sales_invoice'):
		sales_invoice = make_sales_invoice(delivery_note.get('name'))
		sales_invoice.save()
		sales_invoice.submit()

@frappe.whitelist()
def get_items(year):
	fields = ['item_code', 'item_name', 'cannot_be_returned', 'description']
	items = frappe.get_list('Item',
		filters={
			'valid_to_year': ["<=", year],
			'valid_from_year': [">=", year],
			'has_variants': False
		},
		fields=fields,
	)

	items.extend(frappe.get_list('Item',
		filters={
			'valid_to_year': 0,
			'valid_from_year': 0,
			'has_variants': False
		},
		fields=fields,
	))

	for item in items:
		item.qty = 0
		item.uom = 'Nos'
		item.uom_factor = 1
		item.rate = 1 # TODO: Set correct rate.

	return items
